keps/sig-auth/20190408-node-scoped-daemonset.md

```diff
@@ -0,0 +1,239 @@
+to manage the resources in a whole cluster.
+
+**Node-scoped resources**: the resources covered by NodeAuthorizer’s [resource
+graph](https://github.com/kubernetes/kubernetes/blob/4db3fccd6f1c5ae8deb687985a2c8be76e4d4ab5/plugin/pkg/auth/authorizer/node/graph.go#L78),
+currently there are [8
+types](https://github.com/kubernetes/kubernetes/blob/d99f49d2a7caa2e86cb27541a267577771e133a7/plugin/pkg/auth/authorizer/node/graph.go#L123)
+of these resources: node, pod, configmap, pvc, pv, secret, serviceaccount, and
+volumeattachment.
+
+**Node-scoped DaemonSet**: a DaemonSet which restricts a DaemonSet pod to managing
+the node-scoped resources bound to the same node as the DaemonSet pod. A
+node-scoped DaemonSet is still allowed to manage non-node-scoped resources in
+the cluster.
+
+**Node-scoped info**: a boolean type marking whether a DaemonSet should be
+node-scoped or not. For simplicity, the node-scoped info will be at DaemonSet or
+ServiceAccount level, which will apply to both the read and write operations on
+node-scoped resources.
+
+## Proposal
+
+Define the node-scoped info as a new boolean field (`NodeScoped bool`) in
+[ServiceAccountInfo](https://github.com/kubernetes/kubernetes/blob/master/pkg/serviceaccount/util.go#L43),
+and add the node-scoped info into the
+[Extra](https://github.com/kubernetes/kubernetes/blob/255a73ffc64d1c4e0f9a378ebf8d0005edfe6962/staging/src/k8s.io/apiserver/pkg/authentication/user/user.go#L50)
+field of
+[user.DefaultInfo](https://github.com/kubernetes/kubernetes/blob/255a73ffc64d1c4e0f9a378ebf8d0005edfe6962/staging/src/k8s.io/apiserver/pkg/authentication/user/user.go#L46)
+in the
+[ServiceAccountInfo.UserInfo](https://github.com/kubernetes/kubernetes/blob/46c8b075ded386d99de1cf3945168127403c12e9/pkg/serviceaccount/util.go#L48)
+method during authentication. This info, together with the pod name/UID
+information in
+[user.DefaultInfo](https://github.com/kubernetes/kubernetes/blob/255a73ffc64d1c4e0f9a378ebf8d0005edfe6962/staging/src/k8s.io/apiserver/pkg/authentication/user/user.go#L46),
+can then be utilized in NodeAuthorizer and NodeRestriction to implement
+node-scoped DaemonSet.
+
+### The Origin of Node-Scoped Info
+
+Node-scoped info can be defined in two places:
+
+* Option 1: specify the node-scoped info as an annotation (key:
+  `scope.k8s.io/node-scoped`; value: `true` or `false` (default)) or a new
+  boolean field in
+  [ServiceAccount](https://github.com/kubernetes/kubernetes/blob/255a73ffc64d1c4e0f9a378ebf8d0005edfe6962/staging/src/k8s.io/api/core/v1/types.go#L3706).
+    * Pros: this option is easy to implement.
+    * Cons:
+        * this may not work for the cases where a single service account is used
+          to run multiple DaemonSets which have different scope requirements;
+        * this can be used to set the scope of non-daemonset pods, which might
+          be confusing.
+* Option 2: specify the node-scoped info in DaemonSet either in its
```

Comment:
@janetkuo , the node-scoped info can be specified either in ServiceAccount (option 1) or DaemonSet (option 2).
Currently, we favor option 1.